source/ualloc.c

```diff
@@ -27,12 +27,44 @@ extern int printf(const char *, ...);
 
 extern void * volatile mbed_sbrk_ptr;
 
-// Set debug level to 0 until non-allocating printf is available
-const UAllocDebug_t ualloc_debug_level = UALLOC_DEBUG_NONE;//(DEBUG?UALLOC_DEBUG_MAX:UALLOC_DEBUG_NONE);
+/*****************************************************************************************
+UALLOC_DEBUG_LOG is effectively reserved for tracing memory allocations
+Tracing must be enabled via "yotta config":
+{
+    "debug": {
+        "options": {
+            "memory-trace": true
+        }
+    }
+}
+[TODO]: ualloc_debug_level probably needs better control with "yotta config"
+*****************************************************************************************/
+#ifdef YOTTA_CFG_DEBUG_OPTIONS_MEMORY_TRACE
+const UAllocDebug_t ualloc_debug_level = UALLOC_DEBUG_LOG;
+#else
+const UAllocDebug_t ualloc_debug_level = UALLOC_DEBUG_NONE;
+#endif
 
 // Debug characters
 const char ua_chars[] = "NFEWIL";
 
+/*****************************************************************************************
+The purpose of "prevent_tracing" below is twofold:
+
+1. prevent infinite loops (mbed_ualloc_internal -> ualloc_debug -> printf ->
+   mbed_ualloc_internal -> ualloc_debug -> printf...)
+2. prevent a possible scenario when an interrupt occurs while ualloc_debug
+   is printing debug information and the interrupt also calls 'mbed_ualloc' or
+   another memory operation that would result in ualloc_debug being called,
+   which would in turn result in garbled output. By using "prevent_tracing",
+   the output is kept consistent, but the memory operation invoked in the
+   interrupt handler is not logged.
+
+1 above can be prevented by using a logging function that doesn't allocate memory.
+2 above can be prevented by queueing log messages instead of logging them immediately.
+*****************************************************************************************/
+static volatile int prevent_tracing = 0;
+
 #define ualloc_debug(ADBG_LEVEL, fmt, ...) do {                              \
```

Comment:
this should be inside separate trace library, because for example it might be that sometimes user wan't to change endpoint where to print these debug messages (=traces), or change trace format easily without touching ualloc -library.